CUDA implementation

tresreid opened the pull request:
Includes two cuda versions, one with unified memory and one without (v2). 
includes changes to the Makefile to run the cuda version
Minimal changes to openACC and openMP versions (get and print additional timing information).

areinsvo committed:
Add OpenACC version of code

tresreid committed:
cuda implementation

tresreid committed:
recommit

tresreid committed:
update

tresreid committed:
Makefile with option between openMP and CUDA

tresreid committed:
add gitignore

tresreid committed:
stash

tresreid committed:
non working attempt for cuda version

tresreid committed:
working GPU version before cleaning

tresreid committed:
update gpu streams

tresreid committed:
update gpu wait for streams

tresreid committed:
update

tresreid committed:
add cerati

tresreid committed:
pre-cuda adjustment

tresreid committed:
cuda implementation integrated

tresreid committed:
Cuda shared mem. pre texture

tresreid committed:
cuda version 2. shared mem and grid/block sizes fixed

tresreid committed:
fixed prefetching and the timing using cudas clock

tresreid committed:
update

tresreid committed:
CUDA v2 - not unified memory

tresreid committed:
merge update

cerati commented:
I expected only new files, why there are changes to the Makefile and to other implementations?

cerati commented:
I guess the printouts are fine, but you also remove some files. And the Makefile may mean that other versions won't compile?

tresreid committed:
update makefile

tresreid committed:
revert makefile options

tresreid committed:
revert deleted files

tresreid committed:
fix propagate-toz-test.cpp

tresreid committed:
git ignore update

tresreid commented:
I only added some timing printouts for the acc and omp versions. The actual timing method (using the omp timer) will still have to be changed. 
I've added back the accidentally deleted files.
I've removed the additional flags I added in the makefile. I was told to add them to get the best performance for openMP. I guess you mean those options won't work on other architectures?  
Is that everything to fix? 
The other addition to the make file just includes an option to compile cuda. Also the target file includes the compiler and mode names so more than one binary can be made at once. Is that ok?

cerati commented on Makefile:
```diff
@@ -1,33 +1,35 @@
 ifeq ($(MODE),acc)
-CSRCS = propagate-toz-test_OpenACC.cpp
+CSRCS = propagate-toz-test_OpenACC_v5.cpp
 else
```
why do you change version here? and also at lines 24, 28, 32

tresreid replied:
ok. I've taken out the version numbers. However, I was under the impression that those were the lastest, fastest versions. Is that not the case? Those were the versions I had made comparisons to.

cerati replied:
If that is the case than the changes were fine... but I do not know for sure so I prefer to err on the conservative side. I am going to merge the PR, and we can update later. Thanks and sorry for being picky!

tresreid committed:
makefile versions fixed
